---
title: "Example of qPCR Pipeline Usage"
subtitle: "Copy number GAPDH normalised dataset"
author: "Michal Varga"
date: "Last edited: `r format(Sys.time(), '%Y-%m-%d')`"
output: 
  github_document:
    toc: true
    toc_depth: 2
params:
    target_gene: bIFIT1
    primer_set: PS11
    cell_line: MDBK
    housekeeping_gene: bGAPDH
    conditions: !r c('Mock', 'hRSV 1-24', 'bRSV dSH 1-24', 'bIFNa 5-24')
editor_options: 
  markdown: 
    wrap: 72
---

```{r knitr_libraries, include=F}
library(knitr)
library(kableExtra)
library(rmarkdown)
```

# Introduction

intro to qPCR


Normal ddCt:

```{math}
\text{Relative Quantification} = 2^{\Updelta\Updelta \text{Ct}}
```
```{math}
\Updelta\Updelta \text{Ct} = \Updelta \text{Ct}_{\text{Test Samples}}-\Updelta \text{Ct}_{\text{Calibrator Samples}}
```
```{math}
\Updelta \text{Ct}_{\text{Test Samples}} = \text{Ct}_{\text{Target Gene in Tests}}-\text{Ct}_{\text{Reference Gene in Tests}}
```
```{math}
\Updelta \text{Ct}_{\text{Calibrator Samples}} = \text{Ct}_{\text{Target Gene in Calibrator}}-\text{Ct}_{\text{Reference Gene in Calibrator}}
```

Purpose: 

based on scripts:

-   [standard curve](/Scripts/standard_curve.R)
-   [control gene Ct based](/Scripts/standard_curve.R)
-   [Copy number GAPDH normalised dataset](/Scripts/standard_curve.R)

parameters incuded in script:

-   `target_gene:` bIFIT1
-   `primer_set`: PS11
-   `housekeeping_gene:` bGAPDH
-   `cell_line`: MDBK
-   `conditions:` !r c('Mock', 'hRSV 1-24', 'bRSV dSH 1-24', 'bIFNa 5-24')

Link to [stats pipeline]()

These are the required libraries, used in this project:

```{r required_libraries, warning=FALSE, message=FALSE}
library(tidyverse) # for data manipulation
library(data.table) # for fast data reading and writing
library(scales) # for log10 axis on figures
```

# 1. Standard Curve

## Introduction

Some text about what the fuck

obtaining Ct values from known copy numbers

## Data Loading

Load `standard_curves_data.csv` from `/Data` folder, and filter the target gene and primer set used based on the declared parameters in YAML.

```{r data_loading_standard_curve}
standard_curve_data <- fread('../Data/standard_curves_data.csv') %>%
  filter(Target == params$target_gene,
         Primer_set == params$primer_set)
```
```{r data_loading_standard_curve_display, echo=FALSE, warning=FALSE}
standard_curve_data |> kable()
```

## Linear Model Production
### Construct Linear Model

Create a linear model of standard curve by log10 transforming copy number variables.

```{r standard_curve_model}
model_standard_curve <- lm(log10(Copy_number)~Ct, data = standard_curve_data)
```
```{r standard_curve_model_display, echo=FALSE, warning=FALSE}
lm(log10(Copy_number)~Ct, data = standard_curve_data)
```

### Construct a Prediction Based on the Linear Model

1. Construct prediction model (fit, and upper and lower probability intervals) for graph construction.

```{r standard_curve_model_prediction}
prediction_model_standard_curve <- 10^predict(
  model_standard_curve, 
  interval = 'prediction')
```
```{r standard_curve_model_prediction_display, echo=FALSE, warning=FALSE}
10^predict(
  model_standard_curve, 
  interval = 'prediction'
  ) |> kable()
```

2. Join the prediction data to the original dataset.

```{r standard_curve_dataset_bind}
data_combined_standard_curve <- cbind(
  standard_curve_data,
  prediction_model_standard_curve)
```
```{r standard_curve_dataset_bind_display, echo=FALSE, warning=FALSE}
cbind(
  standard_curve_data,
  prediction_model_standard_curve
  ) |> kable()
```


### Calculate Amplification Efficiency

Calculate the amplification efficiency of the primer set based on the equation below. Save the variable into a formated string, that will be used for figure creation.


```{math}
\text{Amplification Efficiency} = 10^{-1/\text{slope}}-1
```

```{r standard_curve_amplification_efficiency}
efficiency_standard_curve <- paste(
  round(
    (10^(-1/ lm(Ct~log10(Copy_number),
                data = standard_curve_data)[[1]][2]) -1)*100,
    digits = 2),
  '% Amplification Efficiency',
  sep = '')
```
```{r standard_curve_amplification_efficiency_display, echo=FALSE, warning=FALSE}
paste(
  round(
    (10^(-1/ lm(Ct~log10(Copy_number), 
                data = standard_curve_data)[[1]][2]) -1)*100, 
    digits = 2), 
  '% Amplification Efficiency', 
  sep = ''
  )
```

## Figure of Standard Curve
### Constructing Figure

1. Initial figure constructed using `ggplot`, with a title that is based on the target gene and primer set that were specified in `params`.

```{r standard_curve_figure_initial_ggplot}
plot_standard_curve <- ggplot(
  data = data_combined_standard_curve,
  mapping = aes(x = Copy_number,
                y = Ct)) +
  geom_point() +
  stat_smooth(method = lm) +
  labs(
    title = paste(
      params$target_gene, '-', 
      params$primer_set, 'Standard Curve', sep = " "),
    y = 'Cycle Threshold',
    x = 'Copy Number')
```

2. The y-axis was set from 0-40, with breaks every 10 points. The x-axis was set to be log10 format. Both were done using the **`scale`** library.

```{r standard_curve_figure_axis}
plot_standard_curve <- plot_standard_curve +
  scale_y_continuous(
    breaks = seq(from = 0, to = 40, by = 10),
    limits = c(0, 40)) +
  scale_x_log10(
    labels = trans_format("log10", 
                          math_format(10^.x)),
    breaks = trans_breaks("log10", 
                          function(x) 10^x)) +
  annotation_logticks(sides='b')
```

3. The figure format was adjuseted based on the `theme_bw` theme.

```{r standard_curve_figure_theme}
plot_standard_curve <- plot_standard_curve +
  theme_bw() +
  theme(
  plot.title = element_text(size = 15, face = 'bold', hjust = 0.5,
                            margin = margin(8, 0, 8, 0)),
  axis.text.y = element_text(angle = 0, size = 9, vjust = 0.5),
  axis.text.x.bottom = element_text(angle = 0, size = 9, vjust = 0.5),
  axis.title.x = element_text(size = 12, face = 'bold', vjust = -0.5, 
                              margin = margin(0, 8, 0, 0)),
  axis.title.y = element_text(size = 12, face='bold', vjust=-0.5,
                              margin = margin(0, 8, 0, 0)),
  aspect.ratio = 1/2)
```

4. Lastly, the previously calucalated primer amplification efficiency was displayed in the top right corner.

```{r standard_curve_figure_annotation}
plot_standard_curve <- plot_standard_curve +
  annotate('text',
           y = 37.5, 
           x = 1000000, 
           label = efficiency_standard_curve, 
           size = 5)
```
```{r standard_curve_plot_display, echo=FALSE, warning=FALSE}
plot_standard_curve
```

### Saving Figure

The figure was saved using the `ggsave` function in the `/Figures` folder. Its name is based on the target gene and primer set that were specified in `params`.

```{r standard_curve_figure_save, message=FALSE}
ggsave(filename = paste(
              paste("plot_standard_curve", 
                    params$target_gene, 
                    params$primer_set,
                    sep = "_"), '.png', sep = ''), 
       plot = plot_standard_curve, 
       device = 'png', 
       path = '../Figures', 
       dpi = 600, 
       height = 12, 
       width = 20, 
       units = 'cm')
```

# 2. Housekeeping Gene Control
## Introduction

Some text about what a housekeeping gene is and how we can use it to modify our extrapolated copy numbers later.

## Data Loading

Load `ct_data.csv` from `/Data` folder, and filter the housekeeping gene and primer set used based on the declared parameters in `params`.

```{r data_loading_housekeeping}
housekeeping_data <- fread('../Data/ct_data.csv') %>% 
  filter(Target == params$housekeeping_gene, 
         Cell_line == params$cell_line,
         Condition %in% params$conditions)
```
```{r data_loading_housekeeping_display, echo=FALSE, warning=FALSE}
housekeeping_data |> kable()
```

## Processing Data

1. An average Ct value is computed for the control condition
```{math}
\text{control\_mean} = \frac{\sum_{i \in \text{Control}} \text{Ct}_i}{n_{\text{Control}}}
```

2. Background (average control Ct values) is removed, and the resulting Ct values are log2 transformed.
```{math}
\log_2(\Delta \text{Ct}) = 2^{- (\text{Ct} - \text{control\_mean})}
```

3. An average of log2 transformed control values is calculated.
```{math}
\text{control\_mean\_log} = \frac{\sum_{i \in \text{Control}} \log_2(\Delta \text{Ct}_i)}{n_{\text{Control}}}
```

4. Values are standardised to the control log2 values, to make the rest of the changes relative to the control values.
```{math}
\text{Value\_normalised} = \frac{\log_2(\Delta \text{Ct})}{\text{control\_mean\_log}}
```


```{r housekeeping_ddct}
ddct_housekeeping_data <- housekeeping_data |> mutate(
  control_mean = mean(
    Ct[Condition == params$conditions[1]],
    na.rm = T),
  log2_dCt = 2^ (- (Ct - control_mean)),
  control_mean_log = mean(
    log2_dCt[Condition == params$conditions[1]],
    na.rm = T),
  Value_normalised = log2_dCt / control_mean_log)
```
```{r housekeeping_ddct_display, echo=FALSE, warning=FALSE}
ddct_housekeeping_data |> kable()
```

## Factorisation

1. The normalised values are aggregated per condition to create average normalised values.

```{r housekeeping_aggregated, warning=FALSE}
aggregated_housekeeping_data <- aggregate(x=ddct_housekeeping_data, 
                               by=list(ddct_housekeeping_data$Condition),
                               FUN = mean)
```
```{r housekeeping_aggregated_display, echo=FALSE, warning=FALSE}
aggregated_housekeeping_data |> kable()
```

2. The values are arranged based on the order of conditions in `params`.

```{r housekeeping_arranged}
arranged_aggregated_housekeeping_data <- aggregated_housekeeping_data %>%
  arrange(match(Group.1, params$conditions))
```
```{r housekeeping_arranged_display, echo=FALSE, warning=FALSE}
arranged_aggregated_housekeeping_data |> kable()
```

3. The average normalised values of housekeeping gene expression are triplicated and saved as a vector. This will be used to adjust the target gene expression values.

```{r housekeeping_factros_vector}
housekeeping_factor_vector <- rep(arranged_aggregated_housekeeping_data$Value_normalised, 
                                  each=3)
```
```{r housekeeping_factros_vector_display, echo=FALSE, warning=FALSE}
housekeeping_factor_vector
```

# 3. Factorised Copy Number Extrapolation

## Introduction

Some text about what the fuck

## Data Loading and Filtering

Load `ct_data.csv` from `/Data` folder, and filter the target gene and primer set used based on the declared parameters in YAML.

```{r data_loading}
target_data <- fread('../Data/copy_number_extrapolation_data.csv') %>%
  filter(Target == params$target_gene,
         Cell_line == params$cell_line,
         Condition %in% params$conditions)
```
```{r data_loading_display, echo=FALSE, warning=FALSE}
target_data |> kable()
```

## Data Wrangling
### Copy Number Extrapolation

The copy numbers were extrapolated from the obtained Ct values based on the model that was generated from the standard curves.

```{r copy_number_extrapolation}
target_copy_number <- target_data |> 
  mutate(Copy_number = 10^predict(model_standard_curve, 
                                  newdata = target_data))
```
```{r copy_number_extrapolation_display, echo=FALSE, warning=FALSE}
target_copy_number |> kable()
```

### Factorisation Based on a Housekeeping Gene Levels

1. To adjust for the relative amounts of housekeeping gene in the different conditions (which should be by definition equal), the extrapolated copy numbers were normalised based on the average normalised values of housekeeping gene expression vector.

```{math}
\text{Copy\_number\_modified} = \frac{\text{Copy\_number}}{\text{Factor}}
```

2. Next, the average control condition copy number was calculated.

```{math}
\text{Control\_mean} = \frac{\sum_{i \in \text{Control}} \text{Copy\_number\_modified}_i}{n_{\text{Control}}}
```

3. The dataset was standardised based on the average control condition copy number, to make the rest of the changes relative to the control values

```{math}
\text{Value\_normalised} = \frac{\text{Copy\_number\_modified}}{\text{Control\_mean}}
```


```{r copy_number_ddct_factorised}
final_target_data <- target_copy_number |> 
  mutate(
    Factor = housekeeping_factor_vector,
    Copy_number_modified = Copy_number / Factor,
    Control_mean = mean(
      Copy_number_modified[Condition == params$conditions[1]], 
      na.rm = T),
    Value_normalised = Copy_number_modified / Control_mean)
```
```{r copy_number_ddct_factorised_display, echo=FALSE, warning=FALSE}
final_target_data |> kable()
```

#### Save data

The processed dataset was saved using the `fwrite` function of `data.table` in the `/Adjusted-Data` folder. Its name is based on the target gene that were specified in `params`.

```{r save_data}
fwrite(final_target_data, 
       paste('../Adjusted-Data/', 
             paste('Analysed', params$target_gene, sep = "_"), 
             '.csv', 
             sep = ''))
```

## Statistics

The statistical analysis was conducted based on the [Statistical Analysis Pipeline](Reports\Templates\Statistics-pipeline.md)

### Visual Assesment

#### Data Distribution by Boxplot

The visual assesment of normal distribution was done by constructing a boxplot.

```{r stats_boxplot}
boxplot(Value_normalised~Condition, final_target_data)
```

This dataset's normality is difficult to assess visualy.

#### Testing Equality of Variance Assumptions

The visual assesment of equality of variance was done by calling `plot` function on a constructed linear model.

```{r stats_variance_plots}
plot(lm(Value_normalised~Condition, final_target_data))
```

**1st and the last plots:** we want symmetrical data about the 0 horizontal line

**2nd plot:** we want residual points to be as close to the predicted line as possible

**3rd plot:** we want for red line to be approx. horizontal

We see equality of variance for all but one condition.

### Mathematical Assesment
#### Assumption of Normality

To test for the normality per condition, the Shapiro-Wilk normality test was performed on each group of values

**p value \> 0.05 means normal distribution**

```{r shapiro.test_single}
shapiro.test(final_target_data$Value_normalised[1:3]) # test all values in one condition
shapiro.test(final_target_data$Value_normalised[4:6]) # test all values in one condition
shapiro.test(final_target_data$Value_normalised[7:9]) # test all values in one condition
shapiro.test(final_target_data$Value_normalised[10:12]) # test all values in one condition
```

The Shapiro-Wilk normality test was also performed on the whole dataset.

```{r shapiro.test_total}
shapiro.test(residuals(lm(Value_normalised~Condition, final_target_data))) # test all values in the whole dataset
```

Each of the conditions individualy confront to normal distribution, however, the dataset as a whole does not meet this assumption.

**Result: NON NORMAL DISTRIBUTION**

#### Assumption of Homogeniety of Variance of Residuals for **Non-Normal Distribution**

**p value \> 0.05 means equal variance**

```{r leveneTest, message=FALSE}
library(car, quietly = TRUE)
leveneTest(Value_normalised~Condition, final_target_data)
```

**Result: EQUAL VARIANCE OF RESIDUALS**

### Statistical Parameters for **Non-Normal Distribution** and **Equal Variance of Residuals**

Kruskal test finds there is any significant difference across the whole dataset. If the p-value is **ABOVE** 0.05 the analysis should be stopped here without comparing groups!

Important parameters: chi-squared and p-value (include in reports)

The dunn test displays both comparison matrix and comparison list of tested groups. P values are the individual p values between group combinations.

```{r kruskal.test}
library(dunn.test, quietly = T)
dunn.test(final_target_data$Value_normalised, 
          final_target_data$Condition, 
          altp=T,
          list=T)
```

**Result: TEST PASSED!**

Group-wise comparison is below:

| **Comparison**            | **Z Statistic**| **p-value**|
|---------------------------|----------------|------------|
| bIFNa 5-24 - Mock         |  1.019049      | 0.3082     |
| bRSV dSH 1-24 - Mock      | -1.019049      | 0.3082     |
| hRSV 1-24 - Mock          |  2.038098      | 0.0415*    |

## Final Figure of Relative Gene Expression Levels
### Constructing Figure


1. Initial figure constructed using ggplot. The control condition was removed as by definition it would be 1, and this is redundant. The data was organised per the order of conditions indicated in `params`.

```{r final_figure_initial_ggplot, warning=F}
plot_normalised_values <- ggplot(final_target_data %>% filter(Condition != params$conditions[1])) +
  aes(x = Value_normalised, 
      y = fct_rev(fct_relevel(Condition, 
                              params$conditions[2], 
                              params$conditions[3],
                              params$conditions[4]))) +
  geom_jitter(shape=17, size=2, width = 0, height = 0.3, color="#131112") +
  stat_summary(fun.y=median, geom="point", size=1.8, color="#A41237") 
```

2. The axis titles were defined and the title is based on the target name, as defined in `params`. The x-axis was transformed into log2.


```{r final_figure_x_axis, warning=FALSE}
plot_normalised_values <- plot_normalised_values +
  labs(
    title = paste("Gene Expression", params$target_gene, sep = " "),
    y = NULL,
    x = 'Fold Change'
  ) +
  scale_x_continuous(trans = log2_trans(),
                     breaks = trans_breaks("log2", function(x) 2^x, n = 8),
                     labels = trans_format("log2", math_format(2^.x)),
                     limits = c(2^-5,2^10),
                     sec.axis = sec_axis(trans = identity,
                                         breaks = c(2^-4, 2^-2, 2^0, 2^2, 2^4, 2^6, 2^8, 2^10),
                                         labels = c(0.062, 0.25, 1, 4, 16, 64, 256, 1024)))
```

3.The calculated p values were added to the corresponding conditions.

```{r final_figure_p_values}
plot_normalised_values <- plot_normalised_values +
  annotate("text", size = 3, fontface = "bold", x = 600, 
           y = params$conditions[2],  label = "0.0415") +
  annotate("text", size = 3, fontface = "bold", x = 600, 
           y = params$conditions[3], label = "0.3082") +
  annotate("text", size = 3, fontface = "bold", x = 600, 
           y = params$conditions[4], label = "0.3082")
```

4. The figure format was adjuseted based on the `theme_classic` theme.

```{r final_figure_theme, warning=F}
plot_normalised_values <- plot_normalised_values +
  theme_classic() +
  theme(
    plot.title = element_text(size = 15, face = 'bold', hjust = 0.5,
                              margin = margin(0, 0, 5, 0)),
    axis.text.y = element_text(angle=0, size=9, vjust=0.2),
    axis.title.x = element_text(size = 12, face='bold', vjust=-0.5,
                                margin = margin(0, 0, 0, 0)),
    axis.title.y = element_blank(),
    axis.text.x=element_text(angle=0, size=9, vjust=0.5),
    axis.ticks.y=element_blank(),
    panel.grid.major.y = element_line(color = "gray86", size = 0.1, linetype = 1),
    legend.position = "none",
    aspect.ratio = 1/2)
```

5. Black line was added that signify the control value. Two red lines were added indicating |2 log2| difference from the mean, surpassing which would deem the gene expression to be biologically significant.

```{r final_figure_annotation_lines}
plot_normalised_values <- plot_normalised_values  +
  geom_vline(xintercept = 1, 
             linetype = "dotted", 
             color = '#2e222f', 
             linewidth = 1) +
  geom_vline(xintercept = 4, 
             linetype = "dotted", 
             color ="#A41237", 
             linewidth = 1) +
  geom_vline(xintercept = 0.25, 
             linetype = "dotted", 
             color = "#A41237", 
             linewidth = 1)
```

```{r final_figure_display, echo=FALSE, warning=FALSE}
plot_normalised_values
```

### Saving Figure

The figure was saved using the `ggsave` function in the `/Figures` folder. Its name is based on the target gene that were specified in `params`.

```{r final_figure_save}
ggsave(filename = paste(
              paste("plot_normalised_values", 
                    params$target_gene,
                    sep = "_"), '.png', sep = ''), 
       plot = plot_normalised_values, 
       device = 'png', 
       path = '../Figures', 
       dpi = 600, 
       height = 16, 
       width = 20, 
       units = 'cm')
```
